packages/sanity/src/core/store/_legacy/history/useTimelineStore.ts

```diff
@@ -0,0 +1,209 @@
+  getSnapshot: () => TimelineState
+  subscribe: (callback: () => void) => () => void
+}
+
+/**
+ * Creates a store which handles the creation of a document Timeline,
+ * TimelineController and also fetches pre-requisite document snapshots.
+ *
+ * `TimelineStore` exposes select TimelineController methods used to query
+ * ranges and fetch more transactions. It can also be used with
+ * `useSyncExternalStore` to subscribe to selected state changes.
+ *
+ * @internal
+ * */
+export function useTimelineStore({
+  documentId,
+  documentType,
+  onError,
+  rev,
+  since,
+}: UseTimelineControllerOpts): TimelineStore {
+  const historyStore = useHistoryStore()
+  const snapshotsSubscriptionRef = useRef<Subscription | null>(null)
+  const timelineStateRef = useRef<TimelineState>(INITIAL_TIMELINE_STATE)
+  const client = useClient(DEFAULT_STUDIO_CLIENT_OPTIONS)
+
+  /**
+   * The mutable TimelineController, used internally
+   */
+  const controller = useMemo(
+    () =>
+      historyStore.getTimelineController({
+        client,
+        documentId,
+        documentType,
+      }),
+    [client, documentId, documentType, historyStore]
+  )
+
+  /**
+   * A BehaviorSubject which can be subscribed by multiple observers, which broadcasts
+   * the the latest state of the (mutable) TimelineController.
+   */
+  const timelineController$ = useMemo(
+    () => new BehaviorSubject<TimelineController>(controller),
+    [controller]
+  )
+
+  /**
+   * Broadcasts the updated state on TimelineController's `handler` updates (as well as on mount / unmount).
+   *
+   * Note that TimelineController triggers updates very frequently when fetching transactions, which it does so
+   * in fairly small batches.
+   *
+   * This will also suspend TimelineController transaction fetching on cleanup, ensuring that fetches don't
+   * continue once the document pane has been closed.
+   */
+  useEffect(() => {
+    controller.setRange(since || null, rev || null)
+    timelineController$.next(controller)
+
+    controller.handler = (err, innerController) => {
+      if (err) {
+        timelineController$.error(err)
+      } else {
+        /**
+         * NOTE: TimelineController requires that you call `setRange` manually whenever its internal
+         * timeline has changed (e.g. has a result of fetched transactions).
+         * Ideally, TimelineController would be updated to handle this automatically for us.
+         * Until then, this workaround waits until the next call stack before calling `setRange`.
+         */
+        setTimeout(() => {
+          innerController.setRange(since || null, rev || null)
+          timelineController$.next(innerController)
+        }, 0)
+      }
+    }
+    controller.resume()
+    return () => controller.suspend()
+  }, [rev, since, controller, timelineController$])
+
+  /**
+   * Fetch document snapshots and update the mutable controller.
+   * Unsubscribes on clean up, preventing double fetches in strict mode.
```

Comment:
Would be interesting to hear @stipsan's view on adding workarounds for this type of strict mode behavior. I feel double fetches in strict mode is not something we should actively fight against because that's how strict mode works by design and for a reason, but I don't feel confident enough about this to have strong opinions.

Reply:
I agree, this adds a lot of noise – and this is more of a quick workaround to prevent `TimelineController` from sporadically crashing due to multiple snapshot fetches. I should have marked it up as such.

At the very least – if we do choose to handle this behaviour – I imagine it's something we'd want to do within fetch helpers / custom hooks and not within consuming components (like the above).

Reply:
I didn't see this, we should never ever fight strict mode 😅 We should refactor this ASAP, and make sure we're not doing it elsewhere. If it's a problem we should instead disable Strict Mode, or only enable strict mode on parts of the render tree that falls outside the crash you're mentioning.

THere's no way to detect if React is doing a double useEffect setup + teardown + setup + teardown, or if there if it's in a background render, transition or a valid race condition 😰

Reply:
For example, this is a more idiomatic way to solve it, that also reduces strict mode noise:
```ts
 useEffect(() => {
    const subscription = remoteSnapshots(
      client,
      {draftId: `drafts.${documentId}`, publishedId: documentId},
      documentType
    ).subscribe((ev: RemoteSnapshotVersionEvent) => {
      controller.handleRemoteMutation(ev)
    })
    return () => {
      subscription.unsubscribe()
    }
  }, [client, controller, documentId, documentType])
```

Where instead of using the "need to fix this thing in strict mode as it's noisy", causing the API to use a `useRef` and more complicated book keeping. The angle is "React might cancel this side effect and trigger clean up at any time and we should anticipate it".

Reply:
Thanks @stipsan. As above, I should have been more explicit in commenting this as a temporary fix. 

The workaround you've suggested will show very subtle side effects in strict mode (you'll see doubled-up entries in the timeline), but I believe this is more an issue with the underlying `TimelineController` and `handleRemoteMutation` not being resilient to handling multiple fetches more than anything.

When time allows – we'll aim to remove that bookkeeping and not fight strict mode as suggested!